src/mpol/coordinates.py

```diff
@@ -50,109 +51,208 @@ class GridCoords:
     :ivar vis_ext: length-4 list of (left, right, bottom, top) expected by routines like ``matplotlib.pyplot.imshow`` in the ``extent`` parameter assuming ``origin='lower'``. Units of [:math:`\mathrm{k}\lambda`]
     """
 
-    def __init__(self, cell_size: float, npix: int) -> None:
-        # set up the bin edges, centers, etc.
-        if not npix % 2 == 0:
-            raise ValueError("Image must have an even number of pixels.")
+    def __init__(self, cell_size: float, npix: int):
+        if npix <= 0 or not (npix % 2 == 0):
+            raise ValueError("Image must have a positive and even number of pixels.")
 
         if cell_size <= 0:
             raise ValueError("cell_size must be a positive real number.")
 
-        self.cell_size = cell_size  # arcsec
-        self.npix = npix
-        self.ncell_u = self.npix
-        self.ncell_v = self.npix
+        # Imply to users that GridCoords instance is read-only and new instance
+        # is the approach if changing values
+        self._cell_size = cell_size
+        self._npix = npix
 
-        # calculate the image extent
-        # say we had 10 pixels representing centers -5, -4, -3, ...
-        # it should go from -5.5 to +4.5
-        lmax = cell_size * (self.npix // 2 - 0.5)
-        lmin = -cell_size * (self.npix // 2 + 0.5)
-        self.img_ext = [lmax, lmin, lmin, lmax]  # arcsecs
+        # Image related
+        self._dimage = cell_size * const.arcsec  # [radians]
+        self._image_centers = self._dimage * (np.arange(npix) - npix // 2)
 
-        self.dl = cell_size * const.arcsec  # [radians]
-        self.dm = cell_size * const.arcsec  # [radians]
+        # Frequency related
+        # These properties are identical for both u & v and defined here
+        self._df = 1 / (npix * self._dimage) * 1e-3  # [kλ]
```

Comment:
I'm not a huge fan of `f` as a variable here. `u` and `v` are commonly used as spatial frequency variables in radio astronomy. `f` implies more about the observing wavelength and would be confusing to users. 

`k` could be a potential variable, but again, it's not really used in this way in radio astronomy. It would be great if you could find a variable that retains its interpretability as either a u or v coordinate.

Reply:
How about `_uv_pixel_width`?

Reply:
Great! But I'm not sure it needs to be a private variable? I could see the user potentially wanting to access this in a few cases.

Reply:
The user will access the hidden variable through the attributes `du` and `dv` which are in`@property` decorators.

Reply:
Oh, yea. Looks good!